File: python/neural_ia/network.py
"""A dependency-free multilayer perceptron for educational experiments."""

from __future__ import annotations

from dataclasses import dataclass
from math import exp, log, sqrt, tanh
from random import Random
from typing import Iterable, Sequence

Sample = tuple[Sequence[float], int]


def _sigmoid(value: float) -> float:
    if value >= 0:
        factor = exp(-value)
        return 1.0 / (1.0 + factor)
    factor = exp(value)
    return factor / (1.0 + factor)


def _activate(name: str, value: float) -> float:
    if name == "tanh":
        return tanh(value)
    if name == "relu":
        return max(0.0, value)
    if name == "sigmoid":
        return _sigmoid(value)
    raise ValueError(f"unsupported activation: {name}")


def _derivative(name: str, output: float) -> float:
    if name == "tanh":
        return 1.0 - output * output
    if name == "relu":
        return 1.0 if output > 0 else 0.0
    if name == "sigmoid":
        return output * (1.0 - output)
    raise ValueError(f"unsupported activation: {name}")


@dataclass
class DenseLayer:
    """A fully connected layer with an element-wise activation."""

    input_size: int
    output_size: int
    activation: str
    weights: list[list[float]]
    biases: list[float]

    @classmethod
    def create(
        cls,
        input_size: int,
        output_size: int,
        activation: str,
        random: Random,
    ) -> "DenseLayer":
        limit = sqrt(6.0 / (input_size + output_size))
        weights = [
            [random.uniform(-limit, limit) for _ in range(input_size)]
            for _ in range(output_size)
        ]
        return cls(input_size, output_size, activation, weights, [0.0] * output_size)

    def forward(self, inputs: Sequence[float]) -> list[float]:
        return [
            _activate(
                self.activation,
                sum(weight * value for weight, value in zip(row, inputs, strict=True))
                + self.biases[index],
            )
            for index, row in enumerate(self.weights)
        ]

    def backward(
        self,
        inputs: Sequence[float],
        outputs: Sequence[float],
        output_gradient: Sequence[float],
        learning_rate: float,
    ) -> list[float]:
        local_gradient = [
            gradient * _derivative(self.activation, output)
            for output, gradient in zip(outputs, output_gradient, strict=True)
        ]
        input_gradient = [0.0] * self.input_size

        for output_index in range(self.output_size):
            for input_index in range(self.input_size):
                input_gradient[input_index] += (
                    local_gradient[output_index] * self.weights[output_index][input_index]
                )

        for output_index in range(self.output_size):
            for input_index in range(self.input_size):
                self.weights[output_index][input_index] -= (
                    learning_rate * local_gradient[output_index] * inputs[input_index]
                )
            self.biases[output_index] -= learning_rate * local_gradient[output_index]

        return input_gradient


class NeuralNetwork:
    """Two-layer MLP for binary classification.

    The implementation intentionally favors readability and deterministic tests
    over tensor-level performance. It mirrors the browser engine used by the
    interactive demo.
    """

    def __init__(self, hidden_size: int = 5, seed: int = 7) -> None:
        if not 2 <= hidden_size <= 16:
            raise ValueError("hidden_size must be between 2 and 16")
        self.hidden_size = hidden_size
        self.seed = seed
        self.epochs = 0
        self._random = Random(seed)
        self.hidden_layer = DenseLayer.create(2, hidden_size, "tanh", self._random)
        self.output_layer = DenseLayer.create(hidden_size, 1, "sigmoid", self._random)

    @staticmethod
    def _validate_input(inputs: Sequence[float]) -> None:
        if len(inputs) != 2:
            raise ValueError("inputs must contain exactly two values")

    def predict(self, inputs: Sequence[float]) -> float:
        self._validate_input(inputs)
        hidden = self.hidden_layer.forward(inputs)
        return self.output_layer.forward(hidden)[0]

    def train_sample(
        self,
        inputs: Sequence[float],
        target: int,
        learning_rate: float = 0.12,
    ) -> float:
        self._validate_input(inputs)
        if target not in (0, 1):
            raise ValueError("target must be 0 or 1")
        if not 0 < learning_rate <= 2:
            raise ValueError("learning_rate must be greater than 0 and at most 2")

        hidden = self.hidden_layer.forward(inputs)
        output = self.output_layer.forward(hidden)
        probability = min(1.0 - 1e-9, max(1e-9, output[0]))
        loss = -(target * log(probability) + (1 - target) * log(1 - probability))
        output_gradient = [
            (probability - target) / (probability * (1.0 - probability))
        ]
        hidden_gradient = self.output_layer.backward(
            hidden, output, output_gradient, learning_rate
        )
        self.hidden_layer.backward(inputs, hidden, hidden_gradient, learning_rate)
        return loss

    def train_epoch(
        self, dataset: Sequence[Sample], learning_rate: float = 0.12
    ) -> float:
        if not dataset:
            raise ValueError("dataset must not be empty")
        order = list(range(len(dataset)))
        self._random.shuffle(order)
        total_loss = 0.0
        for index in order:
            inputs, target = dataset[index]
            total_loss += self.train_sample(inputs, target, learning_rate)
        self.epochs += 1
        return total_loss / len(dataset)

    def fit(
        self,
        dataset: Sequence[Sample],
        epochs: int = 1_500,
        learning_rate: float = 0.12,
    ) -> list[float]:
        if epochs < 1:
            raise ValueError("epochs must be positive")
        return [self.train_epoch(dataset, learning_rate) for _ in range(epochs)]

    def accuracy(self, dataset: Iterable[Sample]) -> float:
        samples = list(dataset)
        if not samples:
            raise ValueError("dataset must not be empty")
        correct = sum(
            (self.predict(inputs) >= 0.5) == bool(target)
            for inputs, target in samples
        )
        return correct / len(samples)

    def to_dict(self) -> dict[str, object]:
        return {
            "format": "neural-ia-python/v1",
            "seed": self.seed,
            "hidden_size": self.hidden_size,
            "epochs": self.epochs,
            "hidden": {
                "weights": self.hidden_layer.weights,
                "biases": self.hidden_layer.biases,
            },
            "output": {
                "weights": self.output_layer.weights,
                "biases": self.output_layer.biases,
            },
        }

File: web/core/network.js
const EPSILON = 1e-9;

export function createRandom(seed = 7) {
  let state = Number(seed) >>> 0;
  return () => {
    state += 0x6d2b79f5;
    let value = state;
    value = Math.imul(value ^ (value >>> 15), value | 1);
    value ^= value + Math.imul(value ^ (value >>> 7), value | 61);
    return ((value ^ (value >>> 14)) >>> 0) / 4294967296;
  };
}

function sigmoid(value) {
  if (value >= 0) {
    const exp = Math.exp(-value);
    return 1 / (1 + exp);
  }

  const exp = Math.exp(value);
  return exp / (1 + exp);
}

function activate(name, value) {
  if (name === "tanh") return Math.tanh(value);
  if (name === "relu") return Math.max(0, value);
  if (name === "sigmoid") return sigmoid(value);
  throw new Error(`Unsupported activation: ${name}`);
}

function derivative(name, output) {
  if (name === "tanh") return 1 - output * output;
  if (name === "relu") return output > 0 ? 1 : 0;
  if (name === "sigmoid") return output * (1 - output);
  throw new Error(`Unsupported activation: ${name}`);
}

class DenseLayer {
  constructor(inputSize, outputSize, activationName, random) {
    this.inputSize = inputSize;
    this.outputSize = outputSize;
    this.activationName = activationName;
    const limit = Math.sqrt(6 / (inputSize + outputSize));
    this.weights = Array.from({ length: outputSize }, () =>
      Array.from({ length: inputSize }, () => (random() * 2 - 1) * limit),
    );
    this.biases = Array(outputSize).fill(0);
  }

  forward(input) {
    return this.weights.map((row, outputIndex) => {
      const weighted = row.reduce(
        (sum, weight, inputIndex) => sum + weight * input[inputIndex],
        this.biases[outputIndex],
      );
      return activate(this.activationName, weighted);
    });
  }

  backward(input, output, outputGradient, learningRate) {
    const localGradient = output.map(
      (value, index) => outputGradient[index] * derivative(this.activationName, value),
    );
    const inputGradient = Array(this.inputSize).fill(0);

    for (let outputIndex = 0; outputIndex < this.outputSize; outputIndex += 1) {
      for (let inputIndex = 0; inputIndex < this.inputSize; inputIndex += 1) {
        inputGradient[inputIndex] +=
          localGradient[outputIndex] * this.weights[outputIndex][inputIndex];
      }
    }

    for (let outputIndex = 0; outputIndex < this.outputSize; outputIndex += 1) {
      for (let inputIndex = 0; inputIndex < this.inputSize; inputIndex += 1) {
        this.weights[outputIndex][inputIndex] -=
          learningRate * localGradient[outputIndex] * input[inputIndex];
      }
      this.biases[outputIndex] -= learningRate * localGradient[outputIndex];
    }

    return inputGradient;
  }
}

export class NeuralNetwork {
  constructor({ hiddenSize = 5, seed = 7, activation = "tanh" } = {}) {
    if (!Number.isInteger(hiddenSize) || hiddenSize < 2 || hiddenSize > 16) {
      throw new RangeError("hiddenSize must be an integer between 2 and 16");
    }

    this.seed = Number(seed) || 7;
    this.hiddenSize = hiddenSize;
    this.activation = activation;
    this.random = createRandom(this.seed);
    this.hiddenLayer = new DenseLayer(2, hiddenSize, activation, this.random);
    this.outputLayer = new DenseLayer(hiddenSize, 1, "sigmoid", this.random);
    this.epochs = 0;
  }

  predict(input) {
    this.#assertInput(input);
    const hidden = this.hiddenLayer.forward(input);
    return this.outputLayer.forward(hidden)[0];
  }

  trainSample(input, target, learningRate = 0.12) {
    this.#assertInput(input);
    if (target !== 0 && target !== 1) {
      throw new RangeError("target must be 0 or 1");
    }
    if (!(learningRate > 0 && learningRate <= 2)) {
      throw new RangeError("learningRate must be greater than 0 and at most 2");
    }

    const hidden = this.hiddenLayer.forward(input);
    const output = this.outputLayer.forward(hidden);
    const probability = Math.min(1 - EPSILON, Math.max(EPSILON, output[0]));
    const loss = -(
      target * Math.log(probability) +
      (1 - target) * Math.log(1 - probability)
    );

    // BCE derivative with respect to sigmoid output. The sigmoid derivative
    // inside DenseLayer reduces this pair to (probability - target).
    const outputGradient = [
      (probability - target) / (probability * (1 - probability)),
    ];
    const hiddenGradient = this.outputLayer.backward(
      hidden,
      output,
      outputGradient,
      learningRate,
    );
    this.hiddenLayer.backward(input, hidden, hiddenGradient, learningRate);
    return loss;
  }

  trainEpoch(dataset, learningRate = 0.12) {
    if (!Array.isArray(dataset) || dataset.length === 0) {
      throw new TypeError("dataset must contain at least one sample");
    }

    const order = Array.from({ length: dataset.length }, (_, index) => index);
    for (let index = order.length - 1; index > 0; index -= 1) {
      const selected = Math.floor(this.random() * (index + 1));
      [order[index], order[selected]] = [order[selected], order[index]];
    }

    let totalLoss = 0;
    for (const index of order) {
      const sample = dataset[index];
      totalLoss += this.trainSample(sample.input, sample.target, learningRate);
    }
    this.epochs += 1;
    return totalLoss / dataset.length;
  }

  snapshot() {
    return {
      seed: this.seed,
      hiddenSize: this.hiddenSize,
      activation: this.activation,
      epochs: this.epochs,
      hiddenWeights: this.hiddenLayer.weights.map((row) => [...row]),
      hiddenBiases: [...this.hiddenLayer.biases],
      outputWeights: this.outputLayer.weights.map((row) => [...row]),
      outputBiases: [...this.outputLayer.biases],
    };
  }

  #assertInput(input) {
    if (
      !Array.isArray(input) ||
      input.length !== 2 ||
      input.some((value) => !Number.isFinite(value))
    ) {
      throw new TypeError("input must be an array with two finite numbers");
    }
  }
}
